File: prolang-audit/python/D_subprograms.py
"""Demonstrate Python subprogram features."""

from functools import reduce


def example_1_first_class_functions() -> None:
    """Store a function inside a variable."""

    def greet(name: str) -> str:
        """Return a greeting."""
        return f"hello, {name}"

    greeter = greet
    # Functions are first-class objects and can be reassigned like any other value.
    print("1. first-class function:", greeter("Python"))


def example_2_closures() -> None:
    """Show a closure capturing outer scope state."""

    def make_multiplier(factor: int):
        """Return a closure that multiplies by factor."""

        def multiply(value: int) -> int:
            """Multiply the incoming value using a captured outer variable."""
            return value * factor

        return multiply

    times_three = make_multiplier(3)
    # The inner function keeps access to factor after the outer call returns.
    print("2. closure:", times_three(5))


def factorial(number: int) -> int:
    """Compute factorial recursively."""
    if number <= 1:
        return 1
    return number * factorial(number - 1)


def example_3_recursion() -> None:
    """Show recursion with factorial."""
    # Python supports direct recursive calls, though depth is bounded by the runtime.
    print("3. recursion:", factorial(5))


def sum_all(*values: int) -> int:
    """Add any number of integer arguments."""
    return sum(values)


def example_4_variadic_parameters() -> None:
    """Show *args variadic parameters."""
    # *args collects positional arguments into a tuple at call time.
    print("4. variadic parameters:", sum_all(1, 2, 3, 4))


def quotient_and_remainder(dividend: int, divisor: int) -> tuple[int, int]:
    """Return two values as a tuple."""
    return dividend // divisor, dividend % divisor


def example_5_multiple_return_values() -> None:
    """Show multiple values returned as a tuple."""
    quotient, remainder = quotient_and_remainder(17, 5)
    # Python packages multiple returns into a tuple and unpacks them naturally.
    print("5. multiple return values:", quotient, remainder)


def example_6_higher_order_functions() -> None:
    """Show map, filter, and reduce as higher-order functions."""
    numbers = [1, 2, 3, 4]
    doubled = list(map(lambda value: value * 2, numbers))
    filtered = list(filter(lambda value: value % 2 == 0, numbers))
    total = reduce(lambda left, right: left + right, numbers)
    # Python passes functions as values into reusable library algorithms.
    print("6. higher-order functions:", doubled, filtered, total)


def example_7_anonymous_functions() -> None:
    """Show a lambda expression."""
    square = lambda value: value * value
    # Lambda provides a compact anonymous function for small expressions.
    print("7. anonymous function:", square(6))


def main() -> None:
    """Run all subprogram examples."""
    example_1_first_class_functions()
    example_2_closures()
    example_3_recursion()
    example_4_variadic_parameters()
    example_5_multiple_return_values()
    example_6_higher_order_functions()
    example_7_anonymous_functions()


if __name__ == "__main__":
    main()

File: prolang-audit/go/d_subprograms.go
package main

import "fmt"

func factorial(number int) int {
	if number <= 1 {
		return 1
	}
	return number * factorial(number-1)
}

func sumAll(values ...int) int {
	total := 0
	for _, value := range values {
		total += value
	}
	return total
}

func quotientAndRemainder(dividend, divisor int) (int, int) {
	return dividend / divisor, dividend % divisor
}

func mapInts(values []int, transform func(int) int) []int {
	result := make([]int, len(values))
	for index, value := range values {
		result[index] = transform(value)
	}
	return result
}

func filterInts(values []int, predicate func(int) bool) []int {
	result := []int{}
	for _, value := range values {
		if predicate(value) {
			result = append(result, value)
		}
	}
	return result
}

func reduceInts(values []int, reducer func(int, int) int, initial int) int {
	total := initial
	for _, value := range values {
		total = reducer(total, value)
	}
	return total
}

func example1FirstClassFunctions() {
	greet := func(name string) string {
		return fmt.Sprintf("hello, %s", name)
	}
	// Function values can be stored in variables and invoked later.
	fmt.Println("1. first-class function:", greet("Go"))
}

func example2Closures() {
	makeMultiplier := func(factor int) func(int) int {
		return func(value int) int {
			return value * factor
		}
	}
	timesThree := makeMultiplier(3)
	// The inner function closes over factor from the outer scope.
	fmt.Println("2. closure:", timesThree(5))
}

func example3Recursion() {
	// Go supports direct recursion with compile-time signature checking.
	fmt.Println("3. recursion:", factorial(5))
}

func example4VariadicParameters() {
	// A trailing ...T parameter collects any number of arguments into a slice.
	fmt.Println("4. variadic parameters:", sumAll(1, 2, 3, 4))
}

func example5MultipleReturnValues() {
	quotient, remainder := quotientAndRemainder(17, 5)
	// Native multi-return is part of the language rather than tuple packing.
	fmt.Println("5. multiple return values:", quotient, remainder)
}

func example6HigherOrderFunctions() {
	numbers := []int{1, 2, 3, 4}
	doubled := mapInts(numbers, func(value int) int { return value * 2 })
	filtered := filterInts(numbers, func(value int) bool { return value%2 == 0 })
	total := reduceInts(numbers, func(left, right int) int { return left + right }, 0)
	// Higher-order behavior comes from passing function literals into helper routines.
	fmt.Println("6. higher-order functions:", doubled, filtered, total)
}

func example7AnonymousFunctions() {
	square := func(value int) int {
		return value * value
	}
	// Anonymous function literals provide lightweight ad hoc behavior.
	fmt.Println("7. anonymous function:", square(6))
}

func main() {
	example1FirstClassFunctions()
	example2Closures()
	example3Recursion()
	example4VariadicParameters()
	example5MultipleReturnValues()
	example6HigherOrderFunctions()
	example7AnonymousFunctions()
}
